connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java

```diff
@@ -0,0 +1,125 @@
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ObjectStorageExecutor {
+
+  private static final Logger log = LoggerFactory.getLogger(ObjectStorageExecutor.class);
+
+  private final Function<DocumentCreationRequest, Document> createDocument;
+  private final Authentication authentication;
+
+  public ObjectStorageExecutor(
+      Authentication authentication, Function<DocumentCreationRequest, Document> createDocument) {
+    this.authentication = authentication;
+    this.createDocument = createDocument;
+  }
+
+  public static ObjectStorageExecutor create(
+      ObjectStorageRequest objectStorageRequest,
+      Function<DocumentCreationRequest, Document> createDocument) {
+    return new ObjectStorageExecutor(objectStorageRequest.getAuthentication(), createDocument);
+  }
+
+  public Object execute(ObjectStorageOperation objectStorageOperation) {
+    return switch (objectStorageOperation) {
+      case DownloadObject downloadObject -> download(downloadObject);
+      case UploadObject uploadObject -> upload(uploadObject);
+    };
+  }
+
+  private UploadResponse upload(UploadObject uploadObject) {
+    String fileName =
+        uploadObject.fileName() != null && !uploadObject.fileName().isEmpty()
+            ? uploadObject.fileName()
+            : uploadObject.document().metadata().getFileName();
+    BlobId blobId = BlobId.of(uploadObject.bucket(), fileName);
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+
+    try {
+      ServiceAccountCredentials sac =
+          ServiceAccountCredentials.fromStream(
+              new ByteArrayInputStream(
+                  authentication.getJsonKey().getBytes(StandardCharsets.UTF_8)));
+      Storage storage =
+          StorageOptions.newBuilder()
+              .setProjectId(uploadObject.project())
+              .setCredentials(sac)
+              .build()
+              .getService();
+
+      Blob blob = storage.createFrom(blobInfo, uploadObject.document().asInputStream());
+      return new UploadResponse(blob.getBucket(), blob.getName());
+    } catch (IOException ioException) {
+      log.error(
+          "GCS: Failed to upload file {} to bucket {}: {}",
+          fileName,
+          uploadObject.bucket(),
+          ioException.getMessage());
+      throw new ConnectorException("Failed to upload file", ioException);
+    }
+  }
+
+  private DownloadResponse download(DownloadObject downloadObject) {
+    StorageOptions storageOptions =
+        StorageOptions.newBuilder().setProjectId(downloadObject.project()).build();
+    try (Storage storage = storageOptions.getService()) {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      BlobId blobId = BlobId.of(downloadObject.bucket(), downloadObject.fileName());
+      storage.downloadTo(blobId, outputStream);
```

Comment:
This puts the file into memory. I recommend checking out the the [`Storage::reader`](https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.Storage#com_google_cloud_storage_Storage_reader_com_google_cloud_storage_BlobId_com_google_cloud_storage_Storage_BlobSourceOption____) method to see if it's possible to achieve something like this:

```
try (ReadChannel reader = storage.reader(blobId)) {
    InputStream inputStream = Channels.newInputStream(reader);
    // now pass the input stream directly to the `DocumentCreationRequest`
}
```